go/mysql/binlog_event_compression.go

```diff
@@ -0,0 +1,254 @@
+//	RAND_EVENT = 13
+//	USER_VAR_EVENT = 14
+//	XID_EVENT = 16
+//	BEGIN_LOAD_QUERY_EVENT = 17
+//	EXECUTE_LOAD_QUERY_EVENT = 18
+//	TABLE_MAP_EVENT = 19
+//	WRITE_ROWS_EVENT_V1 = 23
+//	UPDATE_ROWS_EVENT_V1 = 24
+//	DELETE_ROWS_EVENT_V1 = 25
+//	IGNORABLE_LOG_EVENT = 28
+//	ROWS_QUERY_LOG_EVENT = 29
+//	WRITE_ROWS_EVENT = 30
+//	UPDATE_ROWS_EVENT = 31
+//	DELETE_ROWS_EVENT = 32
+//	XA_PREPARE_LOG_EVENT = 38
+//	PARTIAL_UPDATE_ROWS_EVENT = 39
+//
+// When transaction compression is enabled, the GTID log event has
+// the following fields:
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_payload (1 to N bytes)                |
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_compression_type (1 to 9 bytes)       |
+// +-----------------------------------------+
+// | field_type (1-9 bytes)                  |
+// +-----------------------------------------+
+// | field_size (1-9 bytes)                  |
+// +-----------------------------------------+
+// | m_uncompressed_size size (0 to 9 bytes) |
+// +-----------------------------------------+
+//
+// We need to extract the compressed transaction payload from the GTID
+// event, decompress it with zstd, and then process the internal events
+// (e.g. Query and Row events) that make up the transaction.
+func (ev binlogEvent) TransactionPayload(format BinlogFormat) ([]BinlogEvent, error) {
+	tp := &TransactionPayload{}
+	if err := tp.Decode(ev.Bytes()[format.HeaderLength:]); err != nil {
+		return nil, vterrors.Wrapf(err, "error decoding transaction payload event")
+	}
+	return tp.Events, nil
+}
+
+// Decode decodes and decompresses the payload.
+func (tp *TransactionPayload) Decode(data []byte) error {
+	log.V(5).Infof("Compressed payload event; length: %d, bytes: %v", len(data), data)
+	if err := tp.read(data); err != nil {
+		return err
+	}
+	return tp.decode()
+}
+
+// read unmarshalls the transaction payload event into the
+// TransactionPayload struct. The compressed payload itself will still
+// need to be decoded -- meaning decompressing it and extracting the
+// internal events.
+func (tp *TransactionPayload) read(data []byte) error {
+	pos := uint64(0)
+
+	for {
+		fieldType, ok := readFixedLenUint64(data[pos : pos+1])
+		if !ok {
+			return vterrors.New(vtrpcpb.Code_INTERNAL, "error reading field type")
+		}
+		log.V(5).Infof("Compressed payload event; field type: %d", fieldType)
+		pos++
+
+		if fieldType == payloadHeaderEndMark {
+			tp.Payload = data[pos:]
+			log.V(5).Infof("Compressed payload event; found header end mark; payload as hex: %s", hex.EncodeToString(tp.Payload))
```

Comment:
Similar here, seems like this also could be a huge value. I guess these are mostly some left overs from development, but could we drop them at this point? Looks like there's more logging cases like this but won't comment on each one then as that's a bit redundant.